STYLE.md

```diff
@@ -0,0 +1,432 @@
+
+It is allowed to omit parentheses for only one type class constraint.
+
+If function type signature is very long then place type of each argument under
+its own line with respect to alignment.
+
+```haskell
+putValueInState
+  :: MonadIO m
+  => UserState
+  -> Maybe Int
+  -> AppConfig
+  -> (Int -> m ())
+  -> m ()
+```
+
+If the line with argument names is too big then put each argument on its own line
+and separate it somehow from body section.
+
+```haskell
+putValueInState
+  userState
+  mValue@(Just x)
+  appConfig
+  valueModificator
+  = do
+    <code goes here>
+```
+
+### List Declarations
+
+Align the elements in the list.  Example:
+
+```haskell
+exceptions =
+  [ InvalidStatusCode
+  , MissingContentHeader
+  , InternalServerError
+  ]
+```
+
+### If-then-else clauses
+
+Generally, guards and pattern matches should be preferred over _if-then-else_
+clauses, where possible.  Short cases should usually be put on a single line
+(when line length allows it).
+
+### Case expressions
+
+The alternatives in a case expression can be indented using either of
+the two following styles:
+
+```haskell
+foobar =
+  case something of
+    Just j  -> foo
+    Nothing -> bar
+```
+
+or as
+
+```haskell
+foobar = case something of
+           Just j  -> foo
+           Nothing -> bar
+```
+
+Align the `->` arrows when it helps readability.
+
+Dealing with laziness
+---------------------
+
+By default, use strict data types and lazy functions.
+
+### Data types
+
+Constructor fields should be strict, unless there's an explicit reason
+to make them lazy.  This helps to avoid many common pitfalls caused by too much
+laziness and reduces the number of brain cycles the programmer has to
+spend thinking about evaluation order.
+
+```haskell
+-- Good
+data Point = Point
+  { pointX :: !Double  -- ^ X coordinate
+  , pointY :: !Double  -- ^ Y coordinate
+  }
+```
+
+```haskell
+-- Bad
+data Point = Point
+  { pointX :: Double  -- ^ X coordinate
+  , pointY :: Double  -- ^ Y coordinate
+  }
+```
+
+TODO: Talk to luke
+Additionally, unpacking simple fields often improves performance and
+reduces memory usage:
+
+```haskell
+data Point = Point
+  { pointX :: {-# UNPACK #-} !Double  -- ^ X coordinate
+  , pointY :: {-# UNPACK #-} !Double  -- ^ Y coordinate
+  }
+```
+
+Misc
+----
+
+### Point-free style ###
+
+Avoid over-using point-free style.  For example, this is hard to read:
+
+```haskell
+-- Bad:
+f = (g .) . h
+```
+
+Cabal file formatting
+---------------------
+
+### Modules & libraries
+
+Modules and libraries should go in alphabetical order inside corresponding
+sections. You can put blank lines between groups in each section.
+
+### Warnings ###
+
+Code should be compilable with `-Wall` without warnings.
+
+
+### Default extensions
+
+The only allowed default extension is: `NoImplicitPrelude`
+All other extensions need to be put in the relevant module.
+
+## Extensions to avoid
+
+Avoid `RecordWildCards`, use `NameFieldPuns` instead.
```

Comment:
Would also avoid `ViewPatterns`

Reply:
I'll include this in  the changes after I get some more feedback from others.

Reply:
Updated.

Reply:
Perhaps format these as bullet points.